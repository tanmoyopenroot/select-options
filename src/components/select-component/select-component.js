import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const SelectBoxContainer = styled.div`
  position: relative;
  display: block;
  width: 30%;
  margin: 0 auto;
  font-size: 18px;
  color: #60666d;
`;

const SelectBoxCurrent = styled.div`
  position: relative;
  box-shadow: 0 15px 30px -10px rgba(0, 0, 0, 0.1);
  cursor: pointer;
  outline: none;
`;

const SelectBoxValue = styled.div`
  display: flex;
`;

const SelectBoxValueInput = styled.input`
  display: none;
`;

const SelectBoxValueText = styled.p.attrs({
  display: props => props.show ? 'block' : 'none', 
})`
  display: ${props => props.display};
  width: 100%;
  margin: 0;
  padding: 15px;
  background-color: #fff;
`;

const SelectBoxList = styled.ul`
  position: absolute;
  width: 100%;
  padding: 0;
  list-style: none;
  box-shadow: 0 15px 30px -10px rgba(0,0,0,0.5);
  z-index: 1;
`;

const SelectBoxListOption = styled.label.attrs({
  color: props => props.selected ? 'green' : 'inherit',
  weight: props => props.selected ? 'bold' : 'normal',
})`
  display: block;
  padding: 15px;
  cursor: pointer;
  background-color: #fff;
  color: ${props => props.color};
  font-weight: ${props => props.weight};
`;

export class SelectComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showList: false,
      selectedId: 0,
    };

    this.handleSelect = this.handleSelect.bind(this);
    this.handleToggleList = this.handleToggleList.bind(this);
  }

  handleSelect(id) {
    const { type, setID, saveSelectData } = this.props;
    this.setState({
      selectedId: id,
      showList: true,
    })
    saveSelectData(type, setID, id);
  }

  handleToggleList() {
    const { showList } = this.state;
    this.setState({
      showList: !showList,
    })
  }

  _getList() {
    const { IDList, data, type, setID } = this.props;
    const { selectedId } = this.state;

    return (
      <SelectBoxList>
      {
        IDList.map(key => {
          const { id, name } = data[key];
          return (
            <li
              key={id}
            >
              <SelectBoxListOption
                htmlFor={`set-${setID}-${type}-${id}`}
                onClick={() => this.handleSelect(id)}
                selected={id === selectedId}
              >
                {name}
              </SelectBoxListOption>
            </li>
          )
        })
      }
      </SelectBoxList>
    );
  }

  render () {
    const { IDList, data, type, setID } = this.props;
    const { selectedId } = this.state;

    return (
      <SelectBoxContainer>
        <SelectBoxCurrent
          onClick={this.handleToggleList}
        >
          {
            !selectedId ?
              <SelectBoxValue>
                <SelectBoxValueText
                  show={true}
                >
                  Select Your Option
                </SelectBoxValueText>
              </SelectBoxValue>
            :
              IDList.map(key => {
                const { id, name } = data[key];
                return (
                  <SelectBoxValue
                    key={key}
                  >
                    <SelectBoxValueInput
                      type='radio'
                      id={`set-${setID}-${type}-${id}`}
                      value={id}
                      name={`set-${setID}-${type}`}
                    />
                    <SelectBoxValueText
                      show={id === selectedId}
                    >
                      {name}
                    </SelectBoxValueText>
                  </SelectBoxValue>
                )
              })
          }
        </SelectBoxCurrent>
        {
          this.state.showList ? this._getList() : null
        }
      </SelectBoxContainer>
    );

  }
}

SelectComponent.propTypes = {
  IDList: PropTypes.array,
  data: PropTypes.object,
  type: PropTypes.string,
  setID: PropTypes.number,
  saveSelectData: PropTypes.func,
}

export default SelectComponent;
