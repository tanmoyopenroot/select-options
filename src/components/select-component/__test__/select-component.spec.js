import React from 'react';
import { SelectComponent } from '../select-component';
import { shallow } from 'enzyme';

const setup = () => {
  const users = {
    usersID: [1, 2, 3],
    usersHash: {
      1: { id: 1, name: 'John Doe'}, 
      2: { id: 2, name: 'Alice' },
      3: { id: 3, name: 'Bob' }
    },
  }

  const mocksaveSelectData = jest.fn((type, setID, id) => {});

  return {
    users,
    mocksaveSelectData,
  }
} 

describe('Testing <SelectComponent />', () => {
  const { users, mocksaveSelectData } = setup();
  const { usersID, usersHash } = users;

  const wrapper = shallow(
    <SelectComponent
      IDList={usersID}
      data={usersHash}
      type='user'
      setID={1}
      saveSelectData={mocksaveSelectData}
    />,
  )

  test('Testing SelectComponent\'s props', () => {
    expect(wrapper.length).toEqual(1);
    expect(wrapper.children().length).toBe(1);
    expect(Object.keys(wrapper.instance().props).length).toEqual(5);
    expect(wrapper.instance().props.IDList).toEqual(usersID);
    expect(wrapper.instance().props.data).toEqual(usersHash);
    expect(wrapper.instance().props.type).toEqual('user');
    expect(wrapper.instance().props.setID).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  })

  test('Testing SelectComponent\'s state', () => {  
    expect(wrapper.state().showList).toEqual(false);
    expect(wrapper.state().selectedId).toEqual(0);
  });

  test('Testing SelectComponent\'s functions', () => {  
    wrapper.instance().handleSelect(1)
    expect(mocksaveSelectData.mock.calls.length).toBe(1);
    expect(mocksaveSelectData.mock.calls[0][0]).toBe('user');
    expect(mocksaveSelectData.mock.calls[0][1]).toBe(1);
    expect(mocksaveSelectData.mock.calls[0][2]).toBe(1);
  });
})
