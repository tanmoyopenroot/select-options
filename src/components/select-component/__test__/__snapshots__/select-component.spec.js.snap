// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <SelectComponent /> Testing SelectComponent's props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectComponent
    IDList={
        Array [
            1,
            2,
            3,
          ]
    }
    data={
        Object {
            "1": Object {
              "id": 1,
              "name": "John Doe",
            },
            "2": Object {
              "id": 2,
              "name": "Alice",
            },
            "3": Object {
              "id": 3,
              "name": "Bob",
            },
          }
    }
    saveSelectData={[Function]}
    setID={1}
    type="user"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.div
          onClick={[Function]}
>
          <styled.div>
                    <styled.p
                              show={true}
                    >
                              Select Your Option
                    </styled.p>
          </styled.div>
</styled.div>,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <styled.div>
            <styled.p
                        show={true}
            >
                        Select Your Option
            </styled.p>
</styled.div>,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.p
              show={true}
>
              Select Your Option
</styled.p>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Select Your Option",
              "show": true,
            },
            "ref": null,
            "rendered": "Select Your Option",
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.div
            onClick={[Function]}
>
            <styled.div>
                        <styled.p
                                    show={true}
                        >
                                    Select Your Option
                        </styled.p>
            </styled.div>
</styled.div>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.div>
              <styled.p
                            show={true}
              >
                            Select Your Option
              </styled.p>
</styled.div>,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <styled.p
                show={true}
>
                Select Your Option
</styled.p>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Select Your Option",
                "show": true,
              },
              "ref": null,
              "rendered": "Select Your Option",
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
